compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -706,6 +706,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         expr_ty: Ty<'tcx>,
     ) -> bool {
+        if in_external_macro(self.tcx.sess, expr.span) {
```

Comment:
@compiler-errors 
My change at here makes this case failed:
```rust
rustc --edition 2021 tests/ui/async-await/proper-span-for-type-error.rs
error[E0308]: mismatched types
 --> tests/ui/async-await/proper-span-for-type-error.rs:8:5
  |
8 |     a().await //~ ERROR mismatched types
  |     ^^^^^^^^^ expected enum `Result`, found `()`
  |
  = note:   expected enum `Result<(), i32>`
          found unit type `()`
help: try adding an expression at the end of the block
  |
8 ~     a().await;
9 ~     Ok(()) //~ ERROR mismatched types
  |
```

This is because `in_external_macro` is not right, missing `DesugaringKind::Await`, so I added it and with extra `DesugaringKind::Async`. 

So the new suggestion comes out:
```
_consume_reference::<i32>(&*async { Box::new(7_i32) }.await);
```
It seems right, so I keep it.

By the way, I think our current `in_external_macro` still missing some DesugaringKind, and the `in_external_macro` is not a proper name right now considering it actually not just checking `Macro`, a suitable function name should be something like `proper_for_suggesting`.

And we have a similar function in `span`, but it definitely need some fix:
https://github.com/rust-lang/rust/blob/c04832643d70479988a3ad9d92b63d9bc95e9c6d/compiler/rustc_span/src/lib.rs#L600-L607

We need some cleanup for these functions here?